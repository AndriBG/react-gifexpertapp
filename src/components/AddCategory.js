import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({ setCategories }) => {

    const [ inputValue, setInputValue ] = useState('');

    const handleInputValue = (e) => {
        setInputValue(e.target.value)
    }

    const handleSubmit = (e) =>  {
        e.preventDefault();

        if ( inputValue.trim().length > 1 ) {
            setCategories(catgs => [inputValue, ...catgs]);
            setInputValue('');
        }

    }

    return (
        <form onSubmit= { handleSubmit }>
            {/* <p>{ inputValue }</p> */}
            <input
                type = 'Text'
                placeholder = 'Buscar'
                value = { inputValue }
                onChange = { handleInputValue }
            />
        </form>
    );
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
