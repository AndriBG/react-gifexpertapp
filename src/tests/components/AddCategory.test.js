import React from 'react';
import { shallow } from 'enzyme';
import { AddCategory } from '../../components/AddCategory';

describe('Pruebas en el componente <AddCategory />', () => {

    // 
    const setCategories = jest.fn();

    // se inicializa para tener la ayuda de vscode, pero no es necesario, porque ya se hace en el método "beforeEach"
    let wrapper = shallow(<AddCategory setCategories= { setCategories } />);

    beforeEach( () => {

        // para limpiar si tenemos algún mocks o simulación de algo
        jest.clearAllMocks();

        wrapper = shallow(<AddCategory setCategories= { setCategories } />);

    });

    test('<AddCategory /> debe mostrarse correctamente', () => {

        expect( wrapper ).toMatchSnapshot();

    });

    test('Debe de cambiar la caja(input) de texto', () => {

        const input = wrapper.find('input');
        const value = 'Hola Mundo';

        input.simulate('change', { target: { value } } )

        // Esto es para corroborar que realmente está cambiando el input, y que tiene el value
        expect( wrapper.find('p').text().trim() ).toBe( value );

    });

    test('NO debe de poster la información con submit', () => {

        const form = wrapper.find('form').simulate('submit', {

            preventDefault(){} // hay otra forma de escribirlo, pero esta es más corta.
            // preventDefault: () => {} // Esta es la forma larga

        });

        expect( setCategories ).not.toHaveBeenCalled();

    });

    test('Debe de llamar el setCategories y limpiar la caja de texto(input)', () => {

        const value = 'Hola Mundo';

        wrapper.find('input').simulate('change', { target: { value } });

        wrapper.find('form').simulate('submit', { preventDefault(){} });

        expect( setCategories ).toHaveBeenCalled();

        // Debe ser llamado tres veces
        // expect( setCategories ).toHaveBeenCalledTimes(2);

        // Debe ser llamado pasando como parpametro una función
        expect( setCategories ).toHaveBeenCalledWith( expect.any(Function) );

        expect( wrapper.find('input').prop('value') ).toBe( '' );

    });

});
