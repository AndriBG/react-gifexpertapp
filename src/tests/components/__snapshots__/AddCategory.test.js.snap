// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <AddCategory /> <AddCategory /> debe mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    placeholder="Buscar"
    type="Text"
    value=""
  />
</form>
`;
